src/file/SDL_iostream.c

```diff
@@ -558,6 +558,160 @@ SDL_IOStream *SDL_IOFromFP(FILE *fp, bool autoclose)
+
+// Functions to read/write file descriptors. Not used for windows.
+
+#ifdef HAVE_LSEEK64
+#define lseek_off_t off64_t
+#define lseek       lseek64
+#else
+#define lseek_off_t off_t
+#endif
+
+typedef struct IOStreamFDData
+{
+    int fd;
+    bool autoclose;
+    bool regular_file;
+} IOStreamFDData;
+
+static Sint64 SDLCALL fd_seek(void *userdata, Sint64 offset, SDL_IOWhence whence)
+{
+    IOStreamFDData *iodata = (IOStreamFDData *) userdata;
+    int fdwhence;
+
+    switch (whence) {
+    case SDL_IO_SEEK_SET:
+        fdwhence = SEEK_SET;
+        break;
+    case SDL_IO_SEEK_CUR:
+        fdwhence = SEEK_CUR;
+        break;
+    case SDL_IO_SEEK_END:
+        fdwhence = SEEK_END;
+        break;
+    default:
+        SDL_SetError("Unknown value for 'whence'");
+        return -1;
+    }
+
+	lseek_off_t result = lseek(iodata->fd, (lseek_off_t)offset, fdwhence);
+	if (result < 0) {
+		SDL_SetError("Couldn't get stream offset: %s", strerror(errno));
+	}
+	return result;
+}
+
+static size_t SDLCALL fd_read(void *userdata, void *ptr, size_t size, SDL_IOStatus *status)
+{
+    IOStreamFDData *iodata = (IOStreamFDData *) userdata;
+    ssize_t bytes = read(iodata->fd, ptr, size);
+    if (bytes < 0) {
+        if (errno == EAGAIN) {
+            *status = SDL_IO_STATUS_NOT_READY;
+        } else {
+            SDL_SetError("Error reading from datastream: %s", strerror(errno));
+        }
+		bytes = 0;
+    }
+    return (size_t)bytes;
+}
+
+static size_t SDLCALL fd_write(void *userdata, const void *ptr, size_t size, SDL_IOStatus *status)
+{
+    IOStreamFDData *iodata = (IOStreamFDData *) userdata;
+    ssize_t bytes = write(iodata->fd, ptr, size);
+    if (bytes < 0) {
+        if (errno == EAGAIN) {
+            *status = SDL_IO_STATUS_NOT_READY;
+        } else {
+            SDL_SetError("Error writing to datastream: %s", strerror(errno));
+        }
+		bytes = 0;
+    }
+    return (size_t)bytes;
+}
+
+static bool SDLCALL fd_flush(void *userdata, SDL_IOStatus *status)
+{
+    IOStreamFDData *iodata = (IOStreamFDData *) userdata;
+
+    #if defined(SDL_PLATFORM_APPLE)  // Apple doesn't have fdatasync (rather, the symbol exists as an incompatible system call).
+    if (fcntl(iodata->fd, F_FULLFSYNC) == -1) {
+        return SDL_SetError("Error flushing datastream: %s", strerror(errno));
+    }
+    #elif defined(_POSIX_SYNCHRONIZED_IO)  // POSIX defines this if fdatasync() exists, so we don't need a CMake test.
+        #ifndef SDL_PLATFORM_RISCOS  // !!! FIXME: however, RISCOS doesn't have the symbol...maybe we need to link to an extra library or something?
+        if (iodata->regular_file) {
+            int sync_result;
+#ifdef SDL_PLATFORM_HAIKU
+            sync_result = fsync(iodata->fd);
+#else
+            sync_result = fdatasync(iodata->fd);
+#endif
+            if (sync_result == -1) {
+                return SDL_SetError("Error flushing datastream: %s", strerror(errno));
+            }
+        }
+        #endif
+    #endif
```

Comment:
This all duplicates what happens after `fflush()` in `stdio_flush()`, so it might be better to factor out

```
static bool fd_fdatasync(int fd);
```

and call it with parameter `fileno(iodata->fp)` or `iodata->fd` as appropriate.